Event wip

ylavic opened the pull request:
mpm_event changes/improvements for async, debugging and pass pytests (see each commit's message).

First commits are cherry-picked from @minfrin's work (also in https://github.com/apache/httpd/tree/trunk-ssl-handshake-nonblocking)

ylavic committed:
core,mod_reqtimeout: Add ap_get_connection_timeout().

Provide a new min_connection_timeout hook that modules enforcing a
dynamic connection timeout (e.g. mod_reqtimeout) should use to inform
ap_get_connection_timeout() users about the current timeout being
applied.

Expose the current timeout enforced by mod_reqtimeout by implementing
the min_connection_timeout hook.

* include/ap_mmn.h():
  Minor bump for min_connection_timeout and ap_get_connection_timeout().

* include/http_connection.h():
  Declare min_connection_timeout and ap_get_connection_timeout().

* server/connection.c():
  Implement min_connection_timeout and ap_get_connection_timeout().

* modules/filters/mod_reqtimeout.c(struct reqtimeout_stage_t):
  Add server_timeout as the timeout defined for the server at the current
  stage.

* modules/filters/mod_reqtimeout.c(struct reqtimeout_con_cfg):
  Add time_left as the dynamic timeout enforced by mod_reqtimeout at the
  current stage.

* modules/filters/mod_reqtimeout.c(check_time_left):
  Store the computed time_left in the reqtimeout_con_cfg, and set the
  socket timeout there (returning an error which will be caught if that
  fails).

* modules/filters/mod_reqtimeout.c(extend_timeout):
  Update time_left in the reqtimeout_con_cfg per the time taken by the last
  read.

* modules/filters/mod_reqtimeout.c(reqtimeout_filter):
  Remove the special path for APR_NONBLOCK_READ or AP_MODE_EATCRLF, it
  does the exact same thing than the !(AP_MODE_GETLINE && APR_BLOCK_READ)
  one.

* modules/filters/mod_reqtimeout.c(reqtimeout_init, reqtimeout_before_header,
                                   reqtimeout_before_body, INIT_STAGE):
  Set the server_timeout in the current stage.

* modules/filters/mod_reqtimeout.c(reqtimeout_min_timeout):
  The new hook implementation.

ylavic committed:
mpm_event: Use ap_get_connection_timeout() for CONN_STATE_ASYNC_WAITIO.

If ap_run_process_connection() returns CONN_STATE_ASYNC_WAITIO and the
connection timeout as returned by ap_get_connection_timeout() is different
than the waitio_q timeout, use a timer event rather than the waitio_q to keep
track of the idle connection.

* server/mpm_fdqueue.h(truct timer_event_t):
  Add the "timeout" field to store the timeout of the timer, recomputing
  it from "when" would require to call apr_time_now() otherwise.

* server/mpm/event/event.c():
  #define TIMER_MIN_TIMEOUT as the minimal timer event's timeout, to
  prevent the events from firing before the sockets are added to the
  pollset. Currently set to 50ms (an arbitrary value..).

* server/mpm/event/event.c(struct event_conn_state_t):
  Add the timer_event_t *te field as an alternative to the q.

* server/mpm/event/event.c(struct event_srv_cfg_s):
  Add the server_rec *s field to backref the server_rec and easily pass
  cs->sc->s to ap_get_connection_timeout().

* server/mpm/event/event.c(pollset_add_at, pollset_del_at):
  If the connection is attached to a timer event, log a "t" instead of
  a "q" and the timer's timeout instead of the q's.

* server/mpm/event/event.c(process_socket):
  If ap_get_connection_timeout() is different than the waitio_q timeout,
  acquire a timer event and associate it with the conn_state. A timer
  event associated with a conn_state has a NULL callback (cbfn).

* server/mpm/event/event.c(event_get_timer_event):
  Set the given timeout to the ->timeout field.

* server/mpm/event/event.c(event_register_timed_callback_ex,
                           event_register_poll_callback_ex):
  Return APR_EINVAL if the given callbacks are NULL, this is reserved
  for conn_state timers now. Since it would have crashed at some point
  to pass NULL callbacks before, it's not really an API change.

* server/mpm/event/event.c(listener_thread):
  Fix the poll() timeout set from timers_next_expiry which should be
  taken into account whether it expired or not.
  When a conn_state timer fires/expires, remove it from the pollset and
  abort the connection (with APLOG_INFO).
  When a conn_state timer is polled, cancel the timer.

ylavic committed:
mpm_event: Use r->server's Timeout after the post_read_request hook.

Regardless of keep_alive_timeout_set which anyway is only about the
KeepAliveTimeout to apply _after_ the current request, always use the
request's server Timeout during its processing (i.e. CONN_STATE_HEAR
and CONN_STATE_COMPLETION).

To save the next KeepAliveTimeout to use later, add a new event_srv_cfg
to the conn_state which points to the appropriate server (either r->server
or c->base_server depending on keep_alive_timeout_set as before).

* server/mpm/event/event.c(struct event_conn_state_t):
  Add event_srv_cfg *ka_sc as the server config to apply for kept alive
  connections.

* server/mpm/event/event.c(event_post_read_request):
  Always set cs->sc to the event_srv_cfg or the request's server, and
  point cs->ka_sc to the appropriate one according to keep_alive_timeout_set.

* server/mpm/event/event.c(make_conn_state):
  Initialize cs->ka_sc to the ap_server_conf's event_srv_cfg, like cs->sc.

* server/mpm/event/event.c(process_socket):
  Use cs->ka_sc->ka_q for CONN_STATE_KEEPALIVE.

ylavic committed:
mpm_event: Use monotonic timestamps if available.

If clock_gettime() and CLOCK_MONOTONIC are defined (i.e. most/all?  unixes),
use them to provide a timestamp that never goes past (even if the admin
changes the system time). This avoids entries potentially suddenly expiring
in centuries on a bad clock skew.

* configure.in():
  Provide HAVE_TIME_H, HAVE_CLOCK_GETTIME and HAVE_CLOCK_GETRES.

* server/mpm/event/event.c(event_time_now):
  New helper to get a monotonic timestamp from clock_gettime() if it's
  available, or apr_time_now() (i.e. gettimeofday()) otherwise.

* server/mpm/event/event.c(process_socket, event_resume_suspended,
                           event_get_timer_event, process_lingering_close,
                           listener_thread, event_run):
  Use event_time_now().

covener commented:
Hi @ylavic I have a question about some code that you removed here :)

I am trying to grok current 2.4.x connections_above_limit(). Specifically, why it cares about both threadsperchild and the number of idlers at the same time.  I am hoping you recall what was misguided about it and that will make the rest click for me.

IIUC, the left side is the # of connections that might spring to life and need a thread. We don't want too many to collect into this state, because once they're established it is a pain to shed them.

Other than the flip-flop issue, why did we need to worry about more than threads_per_child*scale on the right side?
Doesn't it double-count the idlers capacity? Or is it some intentional optimism that despite being > tpc*scale we are kind of second-guessing the scaling factor?

I am thinking you came to the same conclusion, I was quite pleased with myself to spot that I arrived (on the surface) at a similar spot as e.g.
https://github.com/apache/httpd/blob/8f8b5ddd607027985792776c81024ba36b3beb61/server/mpm/event/event.c#L4746

But then I took two steps back: in the same WIP PR, i couldn't make much sense of how it works.

Aren't the two sides of pending_above_limit() check not really comparable?  The left side IIUC is the number of idle workers, or the depth of the queue of work for them.  But the limit we compare against is threads_per_child * async_factor?  It seems like both sides need to be scaled (or not) and both sides need to be overall capacity/state or current capacity/state but that kind of gets you back to connections_above_limit but maybe with a different name.

```
    pending_limit = threads_per_child * async_factor;
...
static APR_INLINE int pending_above_limit(int limit)
{
    return ap_queue_info_count(worker_queue_info) < -limit;                                                 
}
...

static APR_INLINE int should_disable_listensocks(void)
{
    if (listensocks_disabled())
        return 0;

    return pending_above_limit(pending_limit);
}
```



```
static APR_INLINE int connections_above_limit(int *busy)
{
    apr_uint32_t i_count = ap_queue_info_num_idlers(worker_queue_info);
    if (i_count > 0) {
        apr_uint32_t c_count = apr_atomic_read32(&connection_count);
        apr_uint32_t l_count = apr_atomic_read32(&lingering_count);
        if (c_count <= l_count
                /* Off by 'listeners_disabled()' to avoid flip flop */
                || c_count - l_count < (apr_uint32_t)threads_per_child +
                                       (i_count - listeners_disabled()) *
                                       (worker_factor / WORKER_FACTOR_SCALE)) {
            return 0;
        }
    }
    else if (busy) {
        *busy = 1;
    }
    return 1;
}
```

ylavic committed:
core: Change filter input/output pending API to return OK/AGAIN/DONE.

Rename the ap_filter_input_pending() and ap_filter_output_pending() core
hooks to ap_core_input_pending() and ap_core_output_pending() respectively.
Change them to return AGAIN when some data are pending.

Add ap_check_input_pending() and ap_check_output_pending() to run the hooks and
fix the return value (DECLINED -> OK, c->aborted -> DONE).

Adapt the callers to the new API (DECLINED -> OK, OK -> AGAIN, DONE).

ylavic committed:
mpm_event: Single linger queue/timeout (short one, 2s).

ylavic committed:
mpm_event: Use atomic reads/writes for shared resources.

ylavic committed:
mpm_event: Periodic linger queue shrink (500ms).

ylavic committed:
mod_status: Be less racy, improve rendering, and show suspended connections.

ylavic committed:
mpm_event: Add kill_connection() to log (APLOG_INFO) interrupted connections.

ylavic committed:
mpm_event: Autotuning from MaxRequestWorkers.

ylavic committed:
core,mpm_event: Non blocking shutdown.

ylavic committed:
mpm_event,listen: Improve/fix children maintenance when num_buckets > 1.

ylavic committed:
mpm_event: Add helpers, simplify code and improve logging before functional changes.

ylavic committed:
mpm_fdqueue: Allow to queue any events (socket, timer, opaque), and use that for mpm_event's backlog queue.

ylavic committed:
mpm_event: Periodic scoreboard stats update (1s).

ylavic committed:
mod_ssl: Nonblocking/async handshakes in CONN_STATE_PROCESSING phase.

If AP_MPMQ_CAN_WAITIO, make mod_ssl perform non blocking TLS handshakes and
go async when it would block.

ylavic committed:
core,http: Non blocking HTTP header read.

ylavic committed:
mpm_event: No need/use of "clogged" connections count, axe.

ylavic committed:
mpm_event: Don't shrink keepalive queue when busy/exiting.

ylavic committed:
mpm_event: Propose some new connections_above_limit() heuristics.

ylavic committed:
mod_proxy,mpm_event: Replace ap_mpm_register_poll_callback*() by ap_mpm_poll_suspended() to avoid races.
